### PR TITLE
Am cloud revenue

### PR DESCRIPTION
**Pull request by shrouti1507**
## Description of the change

> Description here

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/destinations/am/transform.js [hunk 1]
```diff
@@ -443,20 +447,158 @@ function processProductListAction(message) {
...
+  }
+  if (message.properties.quantity) {
+    delete eventClone.properties.quantity;
+  }
+  if (message.properties.revenue_type) {
+    delete eventClone.properties.revenue_type;
+  }
+  toSendEvents.push(eventClone);
+
+  message.properties.products.forEach(product => {
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:473]: check whether there is products array in payload

Comment (prabrishac) [inline v0/destinations/am/transform.js:473]: check if message.properties is there in payload

File: v0/destinations/am/transform.js [hunk 2]
```diff
@@ -443,20 +447,158 @@ function processProductListAction(message) {
...
-  const eventType = message.event.toLowerCase();
-
-  // Why are we skipping Checkout started and Order Updated?
-  // Order Cancelled is not being sent, but we are handling it here
-  if (
-    eventType === EventType.ORDER_CANCELLED.name ||
-    eventType === EventType.ORDER_REFUNDED.name
-  ) {
-    return processProductListAction(message);
+  if (message.properties.revenue) {
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:507]: same check for message.properties

File: v0/destinations/am/transform.js [hunk 3]
```diff
@@ -443,20 +447,158 @@ function processProductListAction(message) {
...
+    delete eventClone.properties.quantity;
   }
-  return [];
+  if (message.properties.revenue_type) {
+    delete eventClone.properties.revenue_type;
+  }
+
+  // product purchased for every product
+  toSendEvents.push(eventClone);
+  message.properties.products.forEach(product => {
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:525]: same check for product array

File: v0/destinations/am/transform.js [hunk 4]
```diff
@@ -443,20 +447,158 @@ function processProductListAction(message) {
...
+    const eventClonePurchaseProduct = JSON.parse(JSON.stringify(message));
+    eventClonePurchaseProduct.event = "Product Purchased";
+    eventClonePurchaseProduct.properties = product;
+    toSendEvents.push(eventClonePurchaseProduct);
+  });
+  // revenue event for every product
+  message.properties.products.forEach(product => {
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:532]: same check for product array

**Review by prabrishac — CHANGES_REQUESTED** [review]

**Review by prabrishac — COMMENTED** [review]
We could define less no. of methods and reuse logic.

File: v0/destinations/am/transform.js [hunk 5]
```diff
@@ -443,20 +447,158 @@ function processProductListAction(message) {
...
+    }
+  }
+  if (RevTrackForOnce.properties.products) {
+    delete RevTrackForOnce.properties.products;
+  }
+  sendEvent.push(RevTrackForOnce);
+  return sendEvent;
+}
+
+function bothConditionTrue(message)
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:573]: Provide meaningful method name.

File: v0/destinations/am/transform.js [hunk 6]
```diff
@@ -443,20 +447,158 @@ function processProductListAction(message) {
...
+  delete ProdOnce.properties.quantity;
+  ProdOnce.properties.products = message.properties.products;
+
+  sendEvent.push(ProdOnce);
+
+  const RevTrackForOnce = JSON.parse(JSON.stringify(message));
+  RevTrackForOnce.event = "Tracking Revenue";
+  RevTrackForOnce.properties.revenue = message.properties.revenue;
+  RevTrackForOnce.properties.price = message.properties.price;
+  RevTrackForOnce.properties.quantity = message.properties.quantity;
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:560]: Do we need these assignments?

**Review by prabrishac — CHANGES_REQUESTED** [review]
Some minor review comments to follow code convention.
Please delete file - transform_mine.js
Add method docs (description) for newly introduced methods.

File: v0/destinations/am/transform.js [hunk 7]
```diff
@@ -37,6 +35,7 @@ const {
 const AMUtils = require("./utils");
 
 const logger = require("../../../logger");
+const { compact } = require("lodash");
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:38]: we should remove it as not used.

File: v0/destinations/am/transform.js [hunk 8]
```diff
@@ -208,14 +195,68 @@ function responseBuilderSimple(
...
+      ) {
+        if (message.properties.revenue)
+          rawPayload.revenue = message.properties.revenue;
+        if (message.properties.price) {
+          rawPayload.price = message.properties.price;
+        }
+        if (message.properties.quantity) {
+          rawPayload.quantity = message.properties.quantity;
+        }
+        if (revenueTypeCalculation > 0) {
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:210]: can we make it `>=0` ?

File: v0/destinations/am/transform.js [hunk 9]
```diff
@@ -208,14 +195,68 @@ function responseBuilderSimple(
     default:
       set(rawPayload, "event_properties", message.properties);
       rawPayload.event_type = evType;
+      if (
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:198]: The code would be more concise, modular and readable, if we move the below code in some other function.

File: v0/destinations/am/transform.js [hunk 10]
```diff
@@ -208,14 +195,68 @@ function responseBuilderSimple(
...
+      if (destination.Config.trackRevenuePerProduct && lengthEvent === 1) {
+        if (message.properties.revenue)
+          rawPayload.revenue = message.properties.revenue;
+        if (message.properties.price) {
+          rawPayload.price = message.properties.price;
+        }
+        if (message.properties.quantity) {
+          rawPayload.quantity = message.properties.quantity;
+        }
+        if (message.properties.revenue > 0) {
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:225]: same as above

File: v0/destinations/am/transform.js [hunk 11]
```diff
@@ -208,14 +195,68 @@ function responseBuilderSimple(
...
+      ) {
+        if (message.properties.revenue)
+          rawPayload.revenue = message.properties.revenue;
+        if (message.properties.price) {
+          rawPayload.price = message.properties.price;
+        }
+        if (message.properties.quantity) {
+          rawPayload.quantity = message.properties.quantity;
+        }
+        if (message.properties.revenue > 0) {
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:247]: same as above

File: v0/destinations/am/transform.js [hunk 12]
```diff
@@ -417,75 +463,133 @@ function processSingleMessage(message, destination) {
...
-
   return responseBuilderSimple(
     groupInfo,
     payloadObjectName,
     message,
     evType,
     mappingConfig[category.name],
-    destination
+    destination,
+    lengthEvent,
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:473]: better name would be numberOfEvents

**Review by prabrishac — CHANGES_REQUESTED** [review]

File: v0/destinations/am/transform.js [hunk 13]
```diff
@@ -105,18 +91,80 @@ function getSessionId(payload) {
...
+          rawPayload.revenueType = "Purchased";
+        } else {
+          rawPayload.revenueType = "Refunded";
+        }
+      }
+      if (numberOfEvents === 1) {
+        // when there is no product array only a single event is sent and
+        // revenue is to be calculated from that only
+        if (message.properties.revenue >= 0) {
+          rawPayload.revenueType = "Purchased";
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:127]: we should also consider the `message.properties.revenue_type || message.properties.revenueType`

File: v0/destinations/am/transform.js [hunk 14]
```diff
@@ -105,18 +91,80 @@ function getSessionId(payload) {
...
+      if (message.properties.price) {
+        rawPayload.price = message.properties.price;
+      }
+      if (message.properties.quantity) {
+        rawPayload.quantity = message.properties.quantity;
+      }
+      if (message.event === "Product Purchased") {
+        // here the revenue of the root is sent through revenueTypeCalculation because
+        // product purchased event does not have this field
+        if (revenueTypeCalculation >= 0) {
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:117]: we should consider price/revenue of each product to determine revenueType for product purchased event

**Review by prabrishac — CHANGES_REQUESTED** [review]

**Review by sayan-mitra — APPROVED** [review]

**Review by prabrishac — APPROVED** [review]

**Pull request merged**